kafka/producer_metric_test.go

```diff
@@ -0,0 +1,143 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build metrics
+// +build metrics
+
+package kafka
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+
+	"github.com/elastic/apm-data/model"
+	apmqueue "github.com/elastic/apm-queue"
+	"github.com/elastic/apm-queue/codec/json"
+)
+
+// run this test with:
+// go test --tags=metrics -run
+// go test -v -run TestProducerMetrics --tags=metrics
+func TestProducerMetrics(t *testing.T) {
+	rdr := sdkmetric.NewManualReader()
+	mp := sdkmetric.NewMeterProvider(sdkmetric.WithReader(rdr))
+
+	otel.SetMeterProvider(mp)
+
+	topic := apmqueue.Topic("default-topic")
+	_, brokers := newClusterWithTopics(t, topic)
+	codec := json.JSON{}
+	producer, err := NewProducer(ProducerConfig{
+		CommonConfig: CommonConfig{
+			Brokers:        brokers,
+			Logger:         zap.NewNop(),
+			TracerProvider: trace.NewNoopTracerProvider(),
+			MeterProvider:  mp,
+		},
+		Sync:    true,
+		Encoder: codec,
+		TopicRouter: func(event model.APMEvent) apmqueue.Topic {
+			return topic
+		},
+	})
+	require.NoError(t, err)
+	require.NotNil(t, producer)
+
+	// NOTE: these tests are flaky
+	// Using go-cmp instead of metricdatatest.AssertEqual allow to bypass
+	// checkes on all metric data fields, which in some cases mey result in
+	// flakiness (i.e. attribute changing value).
+	// But that's not enough to remove all flakiness: as Producer is using 2
+	// meter provider, their order in the rm.ScopeMetrics slice may change,
```

Comment:
We should probably filter out `ScopeMetrics.Scope == instrumentName` (`github.com/elastic/apm-queue/kafka`), then the test won't be flaky.